keps/sig-node/3759-sidecar-containers/README.md

```diff
@@ -0,0 +1,1452 @@
+that is being added to the resource usage. This section assumes it will be added
+by kubelet independently from this formulae computation.
+
+### Exposing Pod Resource requirements
+
+It’s currently not straightforward for users to know the effective resource
+requirements for a pod. The formula for this is `Max ( Max(initContainers), Sum(Containers)) + pod overhead`.
+This is derived from the fact that init
+containers run serially and to completion before non-init containers.  The
+effective request for each resource is then the maximum of the largest request
+for a resource in any init container, and the sum of that resources across all
+non-init containers.
+
+The introduction of in place pod updates of resource requirement in
+[KEP#1287](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources)
+further complicates effective resource requirement calculation as
+`Pod.Spec.Containers[i].Resources` becomes a desired state field and may not
+represent the actual resources in use.  The KEP notes that:
+
+> Schedulers should use the larger of `Spec.Containers[i].Resources` and
+> `Status.ContainerStatuses[i].ResourcesAllocated` when considering available
+> space on a node.
+
+We can introduce `ContainerUse` to represent this value:
+
+```
+ContainerUse(i) = Max(Spec.Containers[i].Resources, Status.ContainerStatuses[i].ResourcesAllocated)
+```
+
+In the absence of KEP 1287, or if the feature is disabled, `ContainerUse` is
+simply:
+
+```
+ContainerUse(i) = Spec.Containers[i].Resources
+```
+
+The sidecar KEP also changes that calculation to be more complicated as sidecar
+containers are init containers that do not terminate.  Since init containers
+start in order, sidecar resource usage needs to be summed into those init
+containers that start after the sidecar.  Defining `InitContainerUse` as:
+
+```
+InitContainerUse(i) = Sum(sidecar containers with index < i) + Max(Spec.InitContainers[i].Resources, Status.InitContainerStatuses[i].ResourcesAllocated)
+```
+
+allows representing the new formula for a pods resource usage
+
+```
+Max ( Max( each InitContainerUse ) , Sum(Sidecar Containers) + Sum(each ContainerUse) ) + pod overhead
+```
+
+Even now, users still sometimes find how a pods effective resource requirements
+are calculated confusing or are just unaware of the formula.  The mitigating
+quality to this is that init container resource requests are usually lower than
+the sum of non-init container resource requests, and can be ignored by users in
+those cases.  Software that requires accurate pod resource requirement
+information (e.g. kube-scheduler, kubelet, autoscalers) don't have that luxury.
+Its too much to ask of users to perform this even more complex calculation
+simply to know the amount of free capacity they need for a given resource to
+allow a pod to schedule.
+
+#### Goals of exposing the Pod.ResourceAllocated field
```

Comment:
As for adding a new field `PodStatus.ResourcesAllocated`, I don't favor it as currently proposed in this KEP. The ResourcesAllocated field is used to describe the resource levels at which each container in the pod has been admitted by a node (a.k.a achieved successful `canAdmitPod(...)`). This is not applicable to unassigned or Pending pods.

IIRC, pending pods that are not scheduled due to lack of resources have their pod conditions updated with the underlying reasons. I'm also struggling to see how adding this field equips users with something actionable (programmatically) that can justify its presence in the pod api.

I do think it is a great idea to expose a few simple helpers that computes a pod's resource requirements and allocations to be consumed by metrics, kubectl describe pod etc, and eliminate any code duplications.

Reply:
We did consider adding a whole new collection, but decided after much debate that it didn't add enough value to carry it's own weight.

Regarding `ResourcesAllocated`, do you object to the name or idea?  E.g. if it was `ResourcesRequested` ?

Reply:
> Regarding `ResourcesAllocated`, do you object to the name or idea? E.g. if it was `ResourcesRequested` ?

We discussed this on today's Sidecar meeting. @smarterclayton had initially concern with introducing this field as there is already a metric for this exposed by scheduler.

But after taking thru scenarios:
- User wants to run `kubectl describe pod` to understand how big of the Node to allocate to schedule a pending pod
- User wants to see a real up to date allocated by the Pod, especially when the new requests are being applied using InPlace Pod Vertical Scaling (metrics wouldn't work for this as they show the "final" result).

we seems to be on the same page regarding the need to have a way to get this data quickly.

@vinaykul suggested maybe we add something specifically for `kubectl describe` command instead of changing the object. But I'm not sure if there are precedence.

One question is whether it is a field on PodStatus or a sub-resource of sort. @thockin I do not see any issues with having the field. Do you see any? Would it be better to have some sort of sub-resource?

Reply:
Subresources have a lot of overhead - I don't see how that's justified for this. This doesn't warrant an abstraction (eg /scale) and doesn't need RBAC isolation (eg /status).

I think a field seems just right

Reply:
> We did consider adding a whole new collection, but decided after much debate that it didn't add enough value to carry it's own weight.

I figured as much. To me, at first glance, a separate sidecar/support containers w/ resource allocations similar to normal containers + superseding restartPolicy {Always, OnFailure, Never} with Init containers -> Sidecar/Support containers -> normal/app containers flow seemed to address the service mesh user story when I walked through it. I know I don't have all the context, but I just had to ask :)


> 
> Regarding `ResourcesAllocated`, do you object to the name or idea? E.g. if it was `ResourcesRequested` ?

I don't favor the name, I was _maybe_ on the idea.

If we end up implementing [pod-level requests+limits KEP](https://github.com/kubernetes/enhancements/pull/1592) for aggregated sidecar container resource management, I want to reserve `ResourcesAllocated` for consistency/congruency with container-level `ResourcesAllocated` [as defined in my KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources#api-changes) so that we can extend in-place resize, if desired, to pod level without causing confusion.

As for the idea, after attending today's sidecar meeting, I have a better handle on the need.  IIUC:
  - To handle scheduling failure due to resource paucity, we need to know say TotalResourceRequirements = max(max(initContainer), sum(Containers)) + podOverhead. (_Resize is not in play_)
  - Metrics needs TotalResourceAllocations = sum(max(Container.Requests, ContainerStatus.ResourcesAllocated))) + podOverhead (_Resize may be in play, scheduler assists InProgress resize_)

If subresource is too heavy for this, kubectl describe / extended PodScheduled condition message could provide this info. But it is not a great choice for programmatic processing where we should respond with objects.

Reply:
@thockin Adding a field `PodSpec.TotalResourcesRequested` seems to address the needs of both scheduling failure and metrics. Let's hold `ResourcesAllocated` in reserve for a possible pod-level requests/limits feature. WDYT?

Reply:
Why spec and not status?  I think spec is a request, which is the pod-level requests+limits KEP, but what we're trying to represent here is the result of  a non-obvious calculation, to save people from needing to do it themselves.

Reply:
Oops! Because I have buttons for eyes 😵‍💫 :) 
It should have been [PodStatus.TotalResourcesRequested](https://github.com/kubernetes/enhancements/pull/3761#discussion_r1093297192) which imho more accurately represents what's needed for the use case of adding nodes to get "insufficient resources/unschedulable" pods going as well as what is being proposed for metrics. I agree we should add it .. other options are not great. Thoughts on my suggested name?

Reply:
LOL.  I don't object to the name, but we all know how good I am at naming things (cough)Ingress.

Reply:
Resolving for now. Name can be decided during the API review of a specific PR